File: packages/usf-ts/src/obsidian.ts
import { promises as fs } from "node:fs";
import path from "node:path";

import { parse, type ParsedSkill } from "./parser.js";

const WIKILINK_RE = /\[\[([^\[\]#|]+)(?:#([^\[\]|]+))?(?:\|[^\[\]]+)?\]\]/g;

export type WikilinkMode = "truncate" | "summary" | "full";

export interface WikilinkStats {
  resolved: number;
  truncated: number;
  skipped: number;
  totalTokens: number;
}

export class Vault {
  constructor(
    public readonly root: string,
    public readonly notes: Map<string, string>,
  ) {}

  async getNoteText(name: string): Promise<string | null> {
    const p = this.notes.get(name);
    if (!p) return null;
    return fs.readFile(p, "utf-8");
  }
}

export async function loadVault(root: string): Promise<Vault> {
  const abs = path.resolve(root);
  const notes = new Map<string, string>();
  await walk(abs, notes);
  return new Vault(abs, notes);
}

async function walk(dir: string, notes: Map<string, string>): Promise<void> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const e of entries) {
    if (e.name.startsWith(".")) continue;
    const full = path.join(dir, e.name);
    if (e.isDirectory()) {
      await walk(full, notes);
    } else if (e.isFile() && e.name.endsWith(".md")) {
      const stem = e.name.replace(/\.md$/, "");
      notes.set(stem, full);
    }
  }
}

export function approxTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

export async function resolveWikilinks(
  vault: Vault,
  text: string,
  opts: { mode?: WikilinkMode; perLinkTokenLimit?: number } = {},
  visited: Set<string> = new Set(),
): Promise<{ text: string; stats: WikilinkStats }> {
  const mode = opts.mode ?? "truncate";
  const limit = opts.perLinkTokenLimit ?? 1000;
  if (!["truncate", "summary", "full"].includes(mode)) {
    throw new Error(`invalid wikilink mode: ${mode}`);
  }
  const stats: WikilinkStats = { resolved: 0, truncated: 0, skipped: 0, totalTokens: 0 };

  // Because JS regex .replace can't await, collect matches and build output sequentially.
  const matches: { start: number; end: number; content: string }[] = [];
  for (const m of text.matchAll(WIKILINK_RE)) {
    const name = m[1].trim();
    const section = m[2]?.trim();
    if (visited.has(name)) {
      stats.skipped += 1;
      matches.push({ start: m.index!, end: m.index! + m[0].length, content: `[[${name}]]` });
      continue;
    }
    const raw = await vault.getNoteText(name);
    if (raw == null) {
      stats.skipped += 1;
      matches.push({ start: m.index!, end: m.index! + m[0].length, content: `[[${name}]]` });
      continue;
    }
    const parsed = parse(raw);
    const body = section ? extractSection(parsed, section) : noteBody(parsed);
    let content = applyMode(parsed.frontmatter, body, mode, limit, stats);
    if (content.includes("[[")) {
      const child = new Set(visited);
      child.add(name);
      const resolved = await resolveWikilinks(vault, content, opts, child);
      content = resolved.text;
      stats.resolved += resolved.stats.resolved;
      stats.truncated += resolved.stats.truncated;
      stats.skipped += resolved.stats.skipped;
      stats.totalTokens += resolved.stats.totalTokens;
    }
    stats.resolved += 1;
    stats.totalTokens += approxTokens(content);
    matches.push({ start: m.index!, end: m.index! + m[0].length, content });
  }
  let out = "";
  let cursor = 0;
  for (const m of matches) {
    out += text.slice(cursor, m.start) + m.content;
    cursor = m.end;
  }
  out += text.slice(cursor);
  return { text: out, stats };
}

function noteBody(parsed: ParsedSkill): string {
  if (Object.keys(parsed.sections).length === 0) {
    return parsed.raw.split(/^---\s*$/m).slice(-1)[0].trim();
  }
  return Object.entries(parsed.sections)
    .map(([name, body]) => `# ${name}\n${body}`)
    .join("\n\n");
}

function extractSection(parsed: ParsedSkill, section: string): string {
  for (const [name, body] of Object.entries(parsed.sections)) {
    if (name.toLowerCase() === section.toLowerCase()) return body;
  }
  return noteBody(parsed);
}

function applyMode(
  frontmatter: Record<string, unknown>,
  body: string,
  mode: WikilinkMode,
  limit: number,
  stats: WikilinkStats,
): string {
  if (mode === "full") return body;
  if (mode === "summary") {
    const s = frontmatter["summary"];
    if (typeof s === "string" && s.trim()) return s.trim();
  }
  const chars = limit * 4;
  if (body.length > chars) {
    stats.truncated += 1;
    return body.slice(0, chars).replace(/\s+$/, "") + "\n\n…[truncated]";
  }
  return body;
}

File: packages/usf-py/usf/obsidian.py
"""Obsidian vault integration — load skills and resolve [[wikilinks]] as memory."""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

from .parser import parse

_WIKILINK_RE = re.compile(r"\[\[([^\[\]#|]+)(?:#([^\[\]|]+))?(?:\|[^\[\]]+)?\]\]")


def approx_tokens(text: str) -> int:
    """Approximate token count (len/4) — avoids a tiktoken dependency."""
    return max(1, len(text) // 4)


@dataclass
class WikilinkStats:
    resolved: int = 0
    truncated: int = 0
    skipped: int = 0
    total_tokens: int = 0


@dataclass
class Vault:
    root: Path
    notes: dict[str, Path] = field(default_factory=dict)

    def get_note_text(self, name: str) -> str | None:
        path = self.notes.get(name)
        if not path:
            return None
        return path.read_text(encoding="utf-8")


def load_vault(root: str | Path) -> Vault:
    root_path = Path(root).resolve()
    notes: dict[str, Path] = {}
    for md in root_path.rglob("*.md"):
        # Ignore Obsidian config/plugin dirs
        if any(part.startswith(".") and part != "." for part in md.relative_to(root_path).parts):
            continue
        notes[md.stem] = md
    return Vault(root=root_path, notes=notes)


def resolve_wikilinks(
    vault: Vault,
    text: str,
    mode: Literal["truncate", "summary", "full"] = "truncate",
    per_link_token_limit: int = 1000,
    _visited: set[str] | None = None,
) -> tuple[str, WikilinkStats]:
    """Expand [[wikilinks]] inline. Returns (expanded_text, stats)."""
    if mode not in ("truncate", "summary", "full"):
        raise ValueError(f"invalid wikilink mode: {mode}")
    stats = WikilinkStats()
    visited = set(_visited or ())

    def replace(m: re.Match) -> str:
        name = m.group(1).strip()
        section = m.group(2).strip() if m.group(2) else None
        if name in visited:
            stats.skipped += 1
            return f"[[{name}]]"  # cycle — leave literal
        raw = vault.get_note_text(name)
        if raw is None:
            stats.skipped += 1
            return f"[[{name}]]"

        parsed = parse(raw)
        # Respect explicit opt-out only when resolving by wikilink in summary/truncate mode.
        # (Notes with usf:false are still referenceable as memory.)
        body = _extract_section(parsed, section) if section else _note_body(parsed)
        content = _apply_mode(parsed.frontmatter, body, mode, per_link_token_limit, stats)
        # Recursive resolution
        if "[[" in content:
            child_visited = visited | {name}
            content, child_stats = resolve_wikilinks(
                vault, content, mode=mode, per_link_token_limit=per_link_token_limit,
                _visited=child_visited,
            )
            stats.resolved += child_stats.resolved
            stats.truncated += child_stats.truncated
            stats.skipped += child_stats.skipped
            stats.total_tokens += child_stats.total_tokens
        stats.resolved += 1
        stats.total_tokens += approx_tokens(content)
        return content

    expanded = _WIKILINK_RE.sub(replace, text)
    return expanded, stats


def _note_body(parsed) -> str:
    """Full markdown body with sections joined."""
    if not parsed.sections:
        return parsed.raw.split("---", 2)[-1].strip() if parsed.frontmatter else parsed.raw
    parts = []
    for name, body in parsed.sections.items():
        parts.append(f"# {name}\n{body}")
    return "\n\n".join(parts)


def _extract_section(parsed, section: str) -> str:
    for name, body in parsed.sections.items():
        if name.lower() == section.lower():
            return body
    return _note_body(parsed)


def _apply_mode(
    frontmatter: dict,
    body: str,
    mode: str,
    per_link_token_limit: int,
    stats: WikilinkStats,
) -> str:
    if mode == "full":
        return body
    if mode == "summary":
        summary = frontmatter.get("summary")
        if isinstance(summary, str) and summary.strip():
            return summary.strip()
        # fall through to truncate
    # truncate
    limit_chars = per_link_token_limit * 4
    if len(body) > limit_chars:
        stats.truncated += 1
        return body[:limit_chars].rstrip() + "\n\n…[truncated]"
    return body
